### DISCUSSION TITLE
Need an envelope follower. Your slew limiter isn't quite doing it!

### DISCUSSION THREAD

**Discussion by elanhickler** [RobinSchmidt/RS-MET#334] [Ideas]
I'm using your slew limiter to create an envelope follower, but it suffers from "gurgly modulation" where, for example, the individual peaks of a sawtooth causes undesirable modulation. I think a simple hold function would work. Rise and then do not fall until x seconds have passed.

**Comment by elanhickler** [comment]
You need an "is idle" or "is silent" check for your slew limiters or you need to provide access to last out value.

**Comment by elanhickler** [comment]
alright gonna subclass.

**Comment by elanhickler** [comment]
tested, it only works  for on/off signals. Do I need to have two slew limiters, one for rise/fall detection, the other to actually rise and fall only after the other one is finished falling? hmmmmmmmmmmmm
```cpp
class EnvelopeFollower
{
public:
	EnvelopeFollower()
	{
		delayTrigger.setTriggerFunction([this]() { doSkip = false; });
	}
	~EnvelopeFollower() = default;

	void setSampleRate(double v)
	{
		slewLimiter.setSampleRate(v);
		delayTrigger.setSampleRate(v);
	}

	double getSample(double v)
	{
		double incomingAbsoluteValue = abs(v);

		bool valueIsRising = incomingAbsoluteValue > lastAbsoluteValue;
		lastAbsoluteValue = incomingAbsoluteValue;

		if (valueIsRising)
		{
			doSkip = true;
			delayTrigger.restart();
			heldAbsoluteValue = std::max(heldAbsoluteValue, incomingAbsoluteValue);
		}
		else if (!doSkip)
		{
			heldAbsoluteValue = incomingAbsoluteValue;
		}

		currentSlewedValue = slewLimiter.getSample(heldAbsoluteValue);

		delayTrigger.incrememt();

		return currentSlewedValue;
	}

	bool isIdle()
	{
		return currentSlewedValue <= 1.e-6;
	}

	void setAttackTime(double v)
	{
		slewLimiter.setAttackTime(v*1000.0);
	}

	void setHoldTime(double v)
	{
		delayTrigger.setSeconds(v);
	}

	void setDecayTime(double v)
	{
		slewLimiter.setReleaseTime(v*1000.0);
	}

	RAPT::rsSlewRateLimiter<double, double> slewLimiter;
	DelayedTrigger delayTrigger;
	bool doSkip = false;
	double heldAbsoluteValue = 0;
	double lastAbsoluteValue = 0;
	double currentSlewedValue = 0;
};
```

**Comment by elanhickler** [comment]
bah, tried a bunch of different things. Can't get a hold function to clean up some of the signal.

robin i need an envelope followerrrrrrrrrrrrrrrrrrrr 😢 

I want to get some nice plucky envelopes if possible, maybe a more advanced transient detector is what I need, or just a good hold function.

**Comment by RobinSchmidt** [comment]
i added a class rsSlewRateLimiterWithHold - but did not yet try it. will do now..

**Comment by RobinSchmidt** [comment]
black: enveloped sawtooth, blue: original envelope, green: detected envelope (with hold set to a length of 2 cycles)
![image](https://user-images.githubusercontent.com/1833099/50275651-2794ec00-0440-11e9-9665-09cddfc66f51.png)
![image](https://user-images.githubusercontent.com/1833099/50275700-44312400-0440-11e9-9b91-05017f22bfeb.png)
hmmm...i think, what you describe as "gurgly" could be related to the Gibbs ripples that make the maximum excursion of each cycle slightly different. in the example, i have taken a sample rate of 2 kHz and a sawtooth frequency of 100.5 Hz which results in a repetitive pattern of 10 cycles. that's probably a 10 Hz gurgle frequency that disappears when the frequency is exactly 100 Hz (or some other exact fraction of the sample rate). 
![image](https://user-images.githubusercontent.com/1833099/50275997-1dbfb880-0441-11e9-9810-3fc422d03082.png)
i think  we should try to roll off the Gibbs ripples (maybe with a Bessel lowpass) before applying the envelope follower. i will try that....

**Comment by RobinSchmidt** [comment]
it seems, with a lowpass in front of the slewrate limiter, i can reasonbaly get rid of the gibbs gurgle (this is without hold):
![image](https://user-images.githubusercontent.com/1833099/50281777-d6412880-0450-11e9-85ac-60c568d5f29f.png)
the envelope still shows modulation with the signal's input frequency (and its harmonics). we could try to get rid of them by another lowpass tuned somewhere below the fundamental of the input. but maybe better would be some sort of sliding maximum extractor. i just wonder, how that could be implemented efficiently (without searching in a circular buffer for a maximum at each sample). but maybe i should just implement it the naive way, see if it gives good results and if so, think about efficiency later

**Comment by RobinSchmidt** [comment]
is this for realtime purposes? because if not, i could use the non-realtime envelope detection code that i have recently written

**Comment by elanhickler** [comment]
realtime. I'm working on FMD and need an envelope follower to open/close the filter.

**Comment by elanhickler** [comment]
is hold time in seconds?

**Comment by RobinSchmidt** [comment]
milliseconds - to be consistent with attack and release parameters

**Comment by elanhickler** [comment]
hmmmmm, I might want to go with a treshold-based envelope trigger. No matter what the settings are, you can't get a plucky envelope, and I've tried yours, Cytomic The Drop and Tone2 FilterBank3 envelope followers. They are all similar in that way.

Edit: But definitely still will use the envelope follower in a larger product where I can have trigger-based and envelope following options.

Edit: I dunno, still trying to decide what to do.

**Comment by elanhickler** [comment]
ok made a decision, envelope follower it is.

**Comment by elanhickler** [comment]
It is possible to get plucky envelopes, you just need a plucky input signal. Not ideal.

I'm playing with Cytomic The Drop and they have a trigger system and a "sense" knob. It seems like a transient detector with a sense amount. Yep, that would be ideal. And it has a switch to go from envelope follower to transient detector.

Edit: Do you have a realtime transient detector lying around I can throw in?

**Comment by RobinSchmidt** [comment]
hmm...not currently in the rs-met codebase. i have some verrrry old legacy code lying around where i did something like that:
https://spl.audio/wp-content/uploads/transient_designer_2_9946_manual.pdf
i could drag it in and adapt/update/make-compatible-with-new

**Comment by RobinSchmidt** [comment]
> maybe i should just implement it the naive way, see if it gives good results and if so, think about efficiency later

i think, i have found something relevant:
https://www.nayuki.io/page/sliding-window-minimum-maximum-algorithm

these links may also be useful:
https://stackoverflow.com/questions/8905525/computing-a-moving-maximum
https://stackoverflow.com/questions/4802038/implement-a-queue-in-which-push-rear-pop-front-and-get-min-are-all-consta

**Comment by elanhickler** [comment]
Yeah I can implement and test the transient detector you come up with, not in a rush though.

**Comment by RobinSchmidt** [comment]
the moving-maximum filter seems to work quite well - the red curve is the moving-max-filtered green curve (with a range/support of one cycle for the filter):
![image](https://user-images.githubusercontent.com/1833099/50388642-eab94580-071c-11e9-8a22-168a7ab241fd.png)

**Comment by RobinSchmidt** [comment]
especially in the constant amplitude section, the modulation of the envelope with the sawtooth frequency is amost completely gone:
![image](https://user-images.githubusercontent.com/1833099/50388681-1e489f80-071e-11e9-8fbd-c52cacd66d77.png)
i've implemented the deque-based algorithm described in the nayuki link. and i actually have an idea how to further reduce the worst-case processing cost to O(log(k))

edit: i'm talking about this point in the article:

> Adding an element and updating the deque takes worst-case Θ(k) time but amortized Θ(1) time.

this thing here:

> Looping while the deque is non-empty and its tail element is less than x, we remove the tail element.

is what takes O(k) in the worst case. but it does not utilize the fact that the deque is in fact sorted, so we can use binary search to find the number of elements we can throw away (at once) instead of just salami slicing them away one by one in a linear loop

edit: after thinking about it some more, it might well be the case that reducing the worst-case cost from O(k) to O(log(k)) by binary search could thwart the amortized cost of O(1) of the current algorithm, so it might be not so advisable after all...hmmm....

**Comment by elanhickler** [comment]
Is this a transient detector you're working on, or an improvement to the envelope follower?

**Comment by RobinSchmidt** [comment]
improvement to the envelope follower. however, the (realtime) transient detection algorithm is based on envelope followers also, so any improvement on envelope detectors may benefit transient detection as well

**Comment by elanhickler** [comment]
nice. Blue line looks perfect.

**Comment by RobinSchmidt** [comment]
haha! yes - because that is the original envelope that i plot for reference. the red is the best detected envelope so far. i can get a slight improvement by applying a moving-max and moving-min and take their average. but recovering the (original) blue envelope would be the holy grail

btw. the green and red ones are below the orginal one because of the initial anti-gibbs-gurgle-lowpass. but that's just a matter of scaling

**Comment by elanhickler** [comment]
ah ok! Don't spend too much time on this, the obvious solution is just to have the plugin with a side chain input for envelope.

**Comment by RobinSchmidt** [comment]
ok - if the use case is such that we have the original envelope signal available, then this would indeed be the cleanest solution. but this stuff here may be useful for dynamics processors

**Comment by RobinSchmidt** [comment]
btw. - do you even think that my diagnosis to attribute the gurgle to gibbs ripples is correct? do your input signals that lead to gurgly sounds expose such gibbs ripples? it's probably less of an issue with natural, acoustic sounds. or maybe it is? not sure, if i should expect pronounced gibbs ripples from recordings of natural sounds. but maybe i should

**Comment by elanhickler** [comment]
I think it would be an issue with any signal. To extract the envelope you need to avoid having it be influenced by the dips between oscillations. Even sinewaves will have that.

**Comment by RobinSchmidt** [comment]
dips between oscillations? you mean the dips that occur in the rectified signal when the sinewave goes through zero?:
![image](https://user-images.githubusercontent.com/1833099/50393637-cd559d00-0757-11e9-8b16-d7aefa793731.png)

(plot created with: https://sagecell.sagemath.org/?z=eJwryMkv0UhMKtYozszTqNDU1FEw0FEwMtAEAGSzBtU=&lang=sage)

this is actually precisely what the moving-max filter addresses (picking the maximum value in some range). but high-frequency sinewaves could pose another problem - their envelope may look modulated (to the eye as well as to the detector) due to the fact that in each cycle, the wave is sampled at different phases (that's not an issue with lower frequency sines because their cycle is sampled densely enough to ensure a sample close enough to the actual maximum in each cycle)

**Comment by RobinSchmidt** [comment]
further improvements (blue: original, green: detected):
![image](https://user-images.githubusercontent.com/1833099/50403113-5098e200-079c-11e9-8c39-8a552dc2deca.png)
![image](https://user-images.githubusercontent.com/1833099/50403255-5cd16f00-079d-11e9-8805-91e3741c7e56.png)
but i cheated a little by manually scaling the detected envelope up to match the original. it can also be seen that the detected envelope is always lagging behind the original one. in a realtime context, one would probably compensate by a look-ahead functionality. if we would want to be really fancy, we could try to avoid such a look-ahead/latency by employing some sort of prediction scheme - but maybe that might be overkill and it would never be perfect anyway (there's no way to predict a sudden change in the envelope, like the border between attack and decay, for example - you could only try to continue the direction of past changes)

**Comment by elanhickler** [comment]
So then all we need for a transient detector is something that sends a trigger when the amplitude envelope travels upward fast enough. Maybe if you have a highpass filter after the envelope extractor and the highpassed signal peak reaches the "sensitivity" threshold, it gets triggered. I think this will also allow for signals that might have multi-level edges (when input signal doesn't reach 0 before another transient)

**Comment by RobinSchmidt** [comment]
this looks really close, i think (i applied another bessel filter to smooth out the steps/jaggies and simulated an appropriate look-ahead by just time-shifting the whole envelope):
![image](https://user-images.githubusercontent.com/1833099/50425296-9d032100-0873-11e9-8ed0-dab3cbf00b2e.png)
![image](https://user-images.githubusercontent.com/1833099/50425307-be640d00-0873-11e9-8c03-f855b2bca545.png)

> all we need for a transient detector is something that sends a trigger

oh - so what you actually want is module that just sends a ping when it detects a transient?

**Comment by elanhickler** [comment]
Well I can integrate the pinging part. I just need a value that goes from false to true at the time (the sample) that the transient is detected (when conditions of a transient are met).

**Comment by elanhickler** [comment]
robin, thanks for the envelope follower. It made my plugin:
![image](https://user-images.githubusercontent.com/9116575/50717525-0853ac80-1045-11e9-8803-2b158bd207be.png)
https://www.soundemote.com/labs/flower-child#s3

**Comment by RobinSchmidt** [comment]
the gui looks cool! how do you use it? the code for the new env-follower is actually still in the prototypes section of the codebase - not in the library?

aha! nice! fun to watch!
https://www.youtube.com/watch?v=tdMpRR4AlQ4

**Comment by elanhickler** [comment]
I have my own version of RS-MET, and I copied the needed code over. Also, we are getting more and more out of sync, you gotta upgrade your JUCE so we can get back in sync. I use the envelope follower to open/close the filter frequency and output volume.

**Comment by RobinSchmidt** [comment]
so, are you using an instance of rsMovingMinMaxFilter? do you have a fixed length for the filter or is it user adjustable? because in its current state, i think it actually may cause problems when the length is not fixed. yes - i need to update juce sometime soon. but at the moment i'm really in the flow working on my modal synthesis algorithm again. i hope, i will be able to produce some first sounds with it in the coming days (i mean: have a working AudioModule available in ToolChain - rendering sounds offline in the experiments code section is something i already did long ago, as you know). time to reap the crops, finally. the algorithm was laying dormant sooo long already. if all goes to plan, this will become a very powerful synthesizer for producing acoustic instrument sounds

**Comment by elanhickler** [comment]
Robin, I am in desperate need of you to finish the fft resynthesis. At least get it to the point of removing/extracting harmonics and rebuilding them at a specific frequency and starting phase. That will allow me to make progress on Orange Tree Samples projects.

Seemed like you were finished or nearly finished with this.

**Comment by RobinSchmidt** [comment]
ok - just one more day on modal stuff then back to fft. almost finished? if you refer to the sinusoidal modeling stuff, that's far from being finished. or do you mean just spectrogram analysis/resynthesis? the basic roundtrip works since years, but when trying to modify things, one typically wants a higher level representation

**Comment by elanhickler** [comment]
You just implemented fft harmonic extraction. I'm assuming now everything is ready for phaselocking. Extract harmonics then rebuild them with fixed frequency while using original amplitude.

If that's finished then upgrade your juce so we can get back in sync.

**Comment by RobinSchmidt** [comment]
ok - the modal stuff is now in a state where i can put it aside for while and may revisit it later. it's just the beginning but it already produces some promising sounds in toolchain (i've added some first patches to the preset repo)

i'll reply on the other thread about phaselocking

**Comment by elanhickler** [comment]
I need a transient detector for Flower Child Filter and the improved envelope follower. Is that all ready to use?

**Comment by RobinSchmidt** [comment]
i'll move the envelope follower code over from the prototypes section to the rapt library tomorrow and make a convenience wrapper class that has all the improvements on board already. ..but i thought, you are already using this code (and presumably made such a convenience wrapper yourself?)

transient detector? you mean, the SPL style transient detector? ...i would have to dig out the old legacy code and marry it with the new stuff (i.e. use SPLs "differential envelope" approach (that i have already implemented somewhere - but would also be straightforward to implement again) - but replace the old envelope detectors with the improved ones. that's not yet ready.

i júst had an idea how to solve this issue: 

> high-frequency sinewaves could pose another problem - their envelope may look modulated (to the eye as well as to the detector) due to the fact that in each cycle, the wave is sampled at different phases (that's not an issue with lower frequency sines because their cycle is sampled densely enough to ensure a sample close enough to the actual maximum in each cycle)

don't look for a maximum value of the samples but instead consider triplets of samples, fit a parabola through each triplet and use the maximum of the parabola maxima ....i'll have to think through the details....

**Comment by elanhickler** [comment]
I've discarded all my edits to be in sync with your library, so I'll need you to make it available in your latest version of RS-MET.

In terms of the transient detector, I want it to trigger a basic ADSR so we are no longer using an envelope follower, but just a nice triggered decaying envelope. I'll have both options available for the user.

**Comment by elanhickler** [comment]
What is the class called for the envelope follower I should use for Flower Child Filter?

When you implement the transient detector, let me know what the class is called.

**Comment by elanhickler** [comment]
I never got to use your envelope follower in Flower Child Filter, I've been using the slew limiter this whole time!

**Comment by RobinSchmidt** [comment]
the code for the min/max filter stuff is dragged over (Filters/Basic/MovingWindowFilters.h/cpp) - but i still need to implement the convenience class. basically, wrap the whole processing chain that i use in my `envelopeFollower()` test/expermination function (with which i have created all the plots in this thread) into a class. i was so busy with the supersaw stuff the last few days

**Comment by elanhickler** [comment]
Alright, no problem! I'll release a flower child filter update with a new filter once you get the envelope follower ready to go.

Also, excited about making this subtractive synth with you. Do you want to do 50/50 on this one? So, no hourly pay on the subtractive synth, royalties only. We will both be bringing a lot to this project. Your supersaw stuff and oscillators and other various RS-MET stuff, my filters, GUI, advertising, product design, and I might ask you to help me in some other areas, like getting your UI library to play better with custom UI stuff, and help me do a few more advanced UI stuff like showing waveform plots.

We could aim for a $40 to $60 price point and aim for average 50 sales a month

**Comment by RobinSchmidt** [comment]
> Also, excited about making this subtractive synth with you. Do you want to do 50/50 on this one? So, no hourly pay on the subtractive synth, royalties only.

yes - ok - i didn't write work-hours for the supersaw work. given that in this case, my code is not the major ingredient (as opposed to snowflake) and you do all the business-stuff, 70/30 would be fine as well

do you think, 50 sales per month is realistic? :-O

**Comment by elanhickler** [comment]
I have no idea, but I'm just throwing out a number so I can see how close I am when we actually sell it. We have over 3000 e-mail subscribers btw, soooo... if 1.5% of our subscriber base makes a purchase a month, that's 50 sales, and we have other outlets to reach people.

The point of the 50/50 split is so that you can have more incentive to create a better product. Like I said, I might need help on a few things.

**Comment by elanhickler** [comment]
Gonna need the envelope following stuff nowwwww!

**Comment by RobinSchmidt** [comment]
ok - i'll take a break now from the waveform morphing

> The point of the 50/50 split is so that you can have more incentive to create a better product. Like I said, I might need help on a few things.

ok - i certainly won't oppose it. :-) that's your call. whatever you consider appropriate within that given range.

**Comment by RobinSchmidt** [comment]
ok - the new convenience class replicating my experimental processing chain from the test function is RAPT::rsEnvelopeFollower2 - but it's still preliminary. some things are still missing and some others need more testing. i will continue tomorrow

**Comment by RobinSchmidt** [comment]
i really like, how the pre-filter makes the env-detection totally invariant with respect to peak-amplitude modulation due to Gibbs ripples (presumably, the reason for the "gurgle"):

no Gibbs ripples (f = 100 Hz, fs = 2 kHz):
![image](https://user-images.githubusercontent.com/1833099/57970859-fb49b100-7986-11e9-9060-5e6ce32a6356.png)

slow Gibbs ripples (f = 100.2 Hz, fs = 2 kHz):
![image](https://user-images.githubusercontent.com/1833099/57970871-17e5e900-7987-11e9-8b4c-d8d9db68600d.png)

fast Gibbs ripples (f = 100.5 Hz, fs = 2 kHz):
![image](https://user-images.githubusercontent.com/1833099/57970888-34822100-7987-11e9-96f3-ff850bf26827.png)

blue: true envelope, green: detected envelope (manually delay-compensated). the detector is totally unfazed by these peak modulations. i still need to figure out, if my ad-hoc formula for the delay works in general (i.e. for all settings and inputs). if so, i can provide a getDelay function based on that formula that client code may use for delay compensation. the code also includes a gain factor for the envelope that i adjusted by eye. i was afraid, the desired factor would depend on how much gibbs ripple is present in the input - but apparently it's not. the same compensation factor (it compensates the gain-loss introduced by the pre-filer) works for all 3 signals. good!

**Comment by RobinSchmidt** [comment]
with this envelope detector in our toolbox, we are now in a good position to create really clean dynamics processors.

**Comment by RobinSchmidt** [comment]
...the parameterization via the expected input freq is a bit weird though - no attack/release parameters (they are computed as function of that freq). we may allow a more standard parametrization, too. ...but then we will have to take care, if we need to change the desired compensations factors that i have adjusted by eye

**Comment by elanhickler** [comment]
How about a transient detector that I can use to simply trigger an attack/decay envelope?

Should I simply try applying a highpass filter to this new envelope follower and have a parameter for threshold? So if the threshold rises fast it will create a detectable signal through the highpass filter. If it rises slow, it will be filtered out.

**Comment by RobinSchmidt** [comment]
i think, for a simple trigger, this code here is total overkill. it has so much code to extract a smooth, high-quality envelope

**Comment by RobinSchmidt** [comment]
for a simple transient detector, it's probably a better idea, i drag over the old SPL style "differential-envelope" code. ...or  maybe write it anew (it's straightforward)

**Comment by RobinSchmidt** [comment]
hmm...unfortunately, my overall gain compensation factor worked only for this specific sawtooth signal. for sine and square, the detected envelope gets too loud - so i removed this arbitrary factor from the code

**Comment by elanhickler** [comment]
I'm getting garbage from your new envelope follower. I set the "attack" control on Flower Child Filter to control the `EnvelopeFollower2` frequency. Lower attack is higher freuqnecy, up to 10,000hz. I checked my input values, they don't seem out of the ordinary.

![image](https://user-images.githubusercontent.com/9116575/58322065-bda8b600-7dd4-11e9-9af4-fcd1af35c3cc.png)

implementation (e2 is `EnvelopeFollower2)`:

![image](https://user-images.githubusercontent.com/9116575/58322366-9999a480-7dd5-11e9-90f7-5f19655b5f88.png)

**Comment by elanhickler** [comment]
solved.
![image](https://user-images.githubusercontent.com/9116575/58322768-c7331d80-7dd6-11e9-9d2e-1fab8db60cd8.png)

**Comment by RobinSchmidt** [comment]
actually, i do initialize it in the constructor:

![image](https://user-images.githubusercontent.com/1833099/58323411-2cf9c680-7e24-11e9-85dd-ccac10528b2e.png)

in pre c++11 times, this was the only way to do it and should still work. ...but it seems like this constructor doesn't get called when i create the env-follower object. ...why....?

**Comment by RobinSchmidt** [comment]
ah - no - you are right! it is rsSlewRateLimiterLinear2. 2!!!! ...fixed. thx

**Comment by elanhickler** [comment]
I like this envelope follower better. It's more smooth, sounds more analog, higher quality. Definitely gives a +1 in proffesional sound for Flower Child Filter. But it really needs Attack/Decay. For Hold, I will throw on a sample & hold on top with a smoothing control.

HOWEVER... A transient detector is still sorely needed. Should be prioritized over adding attack/decay. (I know I still need to pay my bill, tried to do it today and noticed the money didn't get sent from paypal...)

**Comment by elanhickler** [comment]
pattern is: old new old new

http://www.elanhickler.com/transfer/old%20vs%20new%20FCF%20envelope%20follower.mp3

**Comment by RobinSchmidt** [comment]
> I know I still need to pay my bill

actually, after transferring the 400-something to greg's bill, your's dropped to 600-something again. i will have to add some new items from recent work, so maybe it will go to 700-something - but it's still below 1000...so i didn't make an invoice yet - so don't bother paying that sheet just yet...

**Comment by elanhickler** [comment]
What would you recommend so that I can hack together an Attack/Decay speed for the new envelope follower? That way you can make a proper Attack/Decay at your convenience. I'm also going to add midi to FlowerChild Filter to trigger the envelope as I need to get FCF updated sooner rather than later. You can create the transient detector at your convenience.

**Comment by RobinSchmidt** [comment]
> What would you recommend so that I can hack together an Attack/Decay speed for the new envelope follower

it's probably best to do that directly in the class rsEnvelopeFollower2. currently, the overall "speed" of the envelope is parametrized via an expected-input-freq meta-parameter which controls attack, release and the min/max-filter settings all at the same time (with some rules that i found by (quick) trial and error). 

i could expose these 3 settings separately to client code and treat the "setInputFrequency" function as some kind of high-level meta-parameter setting....but then the object's state variables may become somewhat inconsistent (low-level params that don't reflect the high-level param)....maybe i should refactor out a class with the low-level settings only.

in any case, there's no meaningful way to do that from client code side without hacking my code. so it's probably something that i should do. 

edit: maybe there is: subclassing, forgetting about setInputFreq (maybe making it inaccessible by a private override) and implementing setAttack/setRelease/setMinMaxFilterSpeed in the subclass. but it really makes more sense, if i do that in my code

**Comment by RobinSchmidt** [comment]
edit: oh - i just see: it also controls the post-filter cutoff. that makes already 4 low-level params for setInputFreq. i tuned their relationships until i was satisfied with the envelope for my example (enveloped sawtooth) test-signal. but maybe my test-signal was not representative of the totality of inputs that could occur...so maybe further research could be done for how these parameters should be set up, depending on the characteristics of the input signal.....

**Comment by elanhickler** [comment]
Are you working on adding up/down speed for the latest envelope follower? I see commits for "attackDecayFilter"

**Comment by RobinSchmidt** [comment]
this attack-decay filter is just for serving as a simple example modulator for working out polyphonic modulation. i don't want to have to deal with the distractions of the additional complexities of the  breakpoint modulator in this context - and i currently have no other modulators

**Comment by RobinSchmidt** [comment]
it's a filter, that has an adjustable attack/decay shape as its impulse-response:
![image](https://user-images.githubusercontent.com/1833099/61967962-09472e00-afd7-11e9-932e-3f6716643f21.png)
it's actually a sort of spin-off from my modal filters which have an attack-decay enveloped sinusoid as impulse respone - it can be seen as the special case, where the sinusoid has zero frequency

**Comment by elanhickler** [comment]
What's the use case?

**Comment by RobinSchmidt** [comment]
use case? for what? for the attack-decay filter? if so - well - exactly this one here: creating a smooth attack-decay envelope - simply and efficiently. i want to later extend it to allow for a two-stage decay and maybe sustain - we'll see.

**Comment by elanhickler** [comment]
Oh I need a smooth envelope for my synths.

This will be perfect for guitar and plucky sounds. For the two-stage decay we need exponential-ish to linear-ish. That's needed for piano-like sounds as well. This will make synths sound a little more acoustic and pleasing to the ear.

**Comment by RobinSchmidt** [comment]
hmm - a (truly) linear shape is difficult with a filter. (IIR) filters always have some sort of exponentially decaying character to them. what we can do, however, is to mix exponentials with different decay times. i already mentioned that in this old thread:
https://github.com/RobinSchmidt/RS-MET/issues/267
...but i tend to think, that a mix of exponential decays is somehow physically more plausible anyway. my book "the physics of musical instruments" says about piano sounds: "the initial decay rate is several times greater than the final decay rate." ...and speaking about decay *rates* seems to imply an exponential shape. i guess, i should analyze the partial envelopes of the piano sound in my "Tests" repo (https://github.com/RobinSchmidt/RS-MET-Tests/tree/master/Data/Samples)

**Comment by elanhickler** [comment]
Will this smooth A/D envelope feature realtime modulation or loop mode? Well, the problem with realtime modulation + loop mode is calculating the envelope output when the loop restarts between samples.

**Comment by elanhickler** [comment]
Wait, if it's a filter... then... is modulation and loop simply a matter of the input signal and/or controlling the filter? So for example, send the filter an oscillating 1 sample pulse wave, loop time is controlled by a combination of the filter decay slope and the oscillator frequency (they need to approximately match up) and realtime modulation is just a matter of modulating oscillator frequency and filter slopes. Yes?

Or what is the input to the filter for your planned smooth A/D envelope?

**Comment by RobinSchmidt** [comment]
if the input is just an impulse, it will just run through the shape plotted above. once. and if you retrigger (i.e. feed another impulse) during it's still not decayed away, the overlapping shapes will add up (unless you do something against it - which is possible - in fact, i think, we can continuously adjust between add-up and don't-add-up behavior - i think, it's just a matter of subtracting (a 0...1 fraction of) the current output from the input impulse). my plan is to feed in an impulse on note-on plus a "sustain" constant during the note is being held (which doesn't work yet - i need to figure out an appropriate scale factor for that constant - i think, it should be the reciprocal of the filter's DC gain - i need to work out the formula for that). of course - by virtue of being a filter - we may send arbitrary other input signals into it. the question is, what makes most sense musically and leads to the most intuitive control parameters

**Comment by elanhickler** [comment]
I'm pretty sure the envelope follower currently causing my plugin to have regular CPU spikes that make it unusable. Could you check out FlowerChildFilter in my repository and debug?

**Comment by elanhickler** [comment]
........wait, somehow it's not happening anymore. I'll let you know if it comes up again.

**Comment by RobinSchmidt** [comment]
smells like the infamous denormal issue is rearing its ugly head again? actually, this is supposed to be dealt once and for all with on the level of jura::AudioPlugin::processBlock

**Comment by elanhickler** [comment]
Could you throw in hold/attack/decay features in the next few weeks? New flower child filter is almost finished.

**Comment by RobinSchmidt** [comment]
into which class? my attack/decay-filter based envelope that the most recent talk is about? or into the envelope follower?

**Comment by RobinSchmidt** [comment]
ok - i have implemented the sustain for the rsAttackDecayEnvelope which is based on rsAttackDecayFilter (which in turn is based on my rsModalFilterWithAttack with zero frequency). this is what i get - a sort of smooth ADSR - no slope discontinuities at the transitions from attack to decay and from sustain to release:

![image](https://user-images.githubusercontent.com/1833099/86500430-41cbbc00-bd91-11ea-9823-fae417e79b15.png)

![image](https://user-images.githubusercontent.com/1833099/86500571-6d02db00-bd92-11ea-92f2-8b723bc5ba9a.png)

![image](https://user-images.githubusercontent.com/1833099/86500583-755b1600-bd92-11ea-91cf-b4a465bfdbe9.png)

looks like we are almost coming full circle to an ADSR envelope generator - not quite, because as it stands, we will always have R==D - decay and release times are equal. but i actually think, it would be no big deal to lift that restriction. there is a little gotcha though: changing the sustain from zero to some higher value, the location of the peak wanders (it occurs later) and it's height increases. i'm afraid, figuring out compensation formulas for this is a formidable math challenge and perhaps it's not even desirable to compensate this? maybe this coupling is "musical"? dunno. ....aaand it's smoother than standard ADSRs

**Comment by RobinSchmidt** [comment]
>  there is a little gotcha though: changing the sustain from zero to some higher value, the location of the peak wanders (it occurs later) and it's height increases.

here i made a plot to show this effect - attack = 50, decay = 100, sustain = 0.0, 0.2, 0.4, 0.6, 0.8, 1.0

![image](https://user-images.githubusercontent.com/1833099/86500909-cff57180-bd94-11ea-8515-a925a5d557e8.png)

in the black curve, the peak occurs exactly at sample 50 (as ordered by the attack parameter) and has height 1. with increasing sustain, location and height of the peak increases

**Comment by elanhickler** [comment]
ooo nice envlope shapes.

yeah I was talking about the envelope follower

**Comment by RobinSchmidt** [comment]
ahh - ok -  i see. you mean in rsEnvelopeFollower2, you need more flexibility with regard to setting up the embedded rsSlewRateLimiterWithHold? currently, we just have rsEnvelopeFollower2<T>::setInputFrequency which in turn calls updateSmoothingFilters, which in turn sets the embedded slewLimiter object to hard-coded settings (which i determined by some (quick) trial and error). you want more detailed access to these internal parameters?

**Comment by elanhickler** [comment]
Yeah

**Comment by RobinSchmidt** [comment]
ok - i'll look into it tomorrow. looks actually like a trivial task - i think, i'll just make the hard-coded 0.1 and 10.0 factors (attack is 1/10 of a cycle, release is 10 cycles) settable members - and may do the same for the hold parameter, too

**Comment by elanhickler** [comment]
It seems that the rsEnvelopeFollower2 is not going to 0 after getting a signal. This bug only happens if attack time is non-zero.

![image](https://user-images.githubusercontent.com/9116575/94502286-af349080-01b8-11eb-8871-eedfe22a473d.png)

**Comment by elanhickler** [comment]
You haven't made progress on envelope follower 2 so I had the idea to use the AttackDecayFilter to get separate attack/decay time and it's just not behaving. It's as if the modulation goes up and down constantly like a sinewave and the output is sending some extreme values.

Also, don't worry about a hold time feature. I'm going to slap on a tempo-synced sample and hold and call it a day.

Going to see if I can just add a separate attack/decay time on envelope follower 2 now by making those functions locally.

**Comment by elanhickler** [comment]
Alright, I got it kind of working, It's buggy, getting some strange startup values and then some sudden reduction in output when changing timing. I still like having the slew rate limiter hold time control. You should implement attack/decay time. I set the smoothing filters to be the max frequency (min time) of the attack and release time. Works pretty well.

**Comment by elanhickler** [comment]
Ok, I figured out one bug.

![image](https://user-images.githubusercontent.com/9116575/95436217-71e2a800-0908-11eb-927c-294b7244e21b.png)

If the slewLimiter.getSample() function receives zeros, the output stops changing in value so the value is held. This means, as soon as EnvelopeFollower2 stops receiving a signal, the envelope stops decaying. This is made more obvious with less decay time.

**Comment by elanhickler** [comment]
HMMMM... alright I'm including gibbs and smooth knobs on Flower Child Filter for the envelope follower. Lots of fun to play with the frequencies of the minmax and post filter and get gurgling or smooth envelopes.
